# eBay Kleinanzeigen a classifieds section of the German eBay website
# The version of the dataset we are working with is a sample of 50,000 data points

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
# there are 50000 thousand rows and 20 columns, three having missin g value  

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

Changing camelcase to snakecase

In [8]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
num_photos            50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [9]:
#import re
#p = re.compile(r'([a-z])([A-Z])')

#autos.columns = [p.sub(r'\1_\2', i).lower() for i in autos.columns]
#print(autos.columns)

In [10]:
#''''''def make_snake(word):
    new_word =""
    for char in word:
        if char == "_": # keep snake case formatting
            new_word = new_word + char
        elif char.islower():
            new_word = new_word + char
        elif char.isupper():
            if new_word[-2] == "_": # only insert _ if there are at least 2 chars after one _
                new_word = new_word + char.lower()
            else:
                new_word = new_word + "_" + char.lower()
    return str(new_word)
#make_snake("dateCrawled")''''''

IndentationError: unexpected indent (<ipython-input-10-fddec7964f2b>, line 2)

In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Initial Data Exploration and Cleaning

In [12]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

##### Any columns that have mostly one value that are candidates to be dropped
###### Any columns that need more investigation.
#### ####Any examples of numeric data stored as text  need to be cleaned.
our observation seller ,offer_type,num_photos column does not have any value and need further investigation 

In [13]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [14]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [15]:
type(autos)

pandas.core.frame.DataFrame

In [16]:
autos.shape

(50000, 17)

In [17]:
print(autos.dtypes)

date_crawled          object
name                  object
price                 object
ab_test               object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer              object
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object


there are some columns price and odometer represented as numric in string format which need cleaning 

In [18]:
autos.loc[:,"price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [19]:
autos.loc[:,"odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [20]:
autos["price"]= autos["price"].str.replace('"','')

In [21]:
print(autos["price"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [22]:
print(autos["price"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [23]:
autos["price"]= autos["price"].str.replace(",","")

In [24]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [25]:
#print(autos["price"].unique())

In [26]:
#autos["price"]= autos["price"].astype(int)

In [27]:
#print(autos["price"].unique())

In [28]:
#print(autos["price"].dtype)#

In [29]:
#autos["price"]= autos["price"].str.replace('$','')

In [30]:
#autos["price"].head(20)

In [31]:
#autos["price"].unique()

In [32]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [33]:
#autos["odometer"] = autos["odometer"].str.replace('"','')

In [34]:
#autos["odometer"] = autos["odometer"].str.replace(',','')

In [35]:
#autos["odometer"] = autos["odometer"].str.replace('km','')

In [36]:
#print(autos["odometer"].unique())

In [37]:
#autos["odometer"] = autos["odometer"].astype(int)

In [38]:
#print(autos["price"].unique())

In [39]:
#autos["odometer"].head()

In [40]:
#autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)

In [41]:
#autos.rename({"date_created":"ad_created"}, axis = 1, inplace = True)

In [42]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
price                 50000 non-null int64
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null int64
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(6), object(11)
memory usage: 6.5+ MB


In [43]:
autos["odometer_km"].shape

(50000,)

In [44]:
autos["price"].shape

(50000,)

In [45]:
autos[["odometer_km","price"]].shape

(50000, 2)

In [46]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [47]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [48]:
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [49]:
autos["odometer_km"].sort_index(ascending = False).head()

49999    150000
49998     40000
49997      5000
49996    150000
49995    100000
Name: odometer_km, dtype: int64

In [50]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

In [51]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [52]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Exploring the date columns
There are a number of columns with date information:
date_crawled
registration_month
registration_year
ad_created
last_seen

In [53]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [54]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          48565 non-null object
name                  48565 non-null object
price                 48565 non-null int64
ab_test               48565 non-null object
vehicle_type          43979 non-null object
registration_year     48565 non-null int64
gearbox               46222 non-null object
power_ps              48565 non-null int64
model                 46107 non-null object
odometer_km           48565 non-null int64
registration_month    48565 non-null int64
fuel_type             44535 non-null object
brand                 48565 non-null object
unrepaired_damage     39464 non-null object
ad_created            48565 non-null object
postal_code           48565 non-null int64
last_seen             48565 non-null object
dtypes: int64(6), object(11)
memory usage: 6.7+ MB


In [55]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().head(20)


(2346,)
count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64


500     781
1500    734
2500    643
1200    639
1000    639
600     531
3500    498
800     498
2000    460
999     434
750     433
900     420
650     419
850     410
700     395
4500    394
300     384
2200    382
950     379
1100    376
Name: price, dtype: int64

In [56]:
autos['last_seen'].value_counts(normalize=True, dropna=False)

2016-04-07 06:17:27    0.000165
2016-04-07 03:16:17    0.000144
2016-04-06 21:17:51    0.000144
2016-04-06 06:17:24    0.000124
2016-04-06 07:46:11    0.000124
2016-04-05 16:44:47    0.000124
2016-04-06 15:17:56    0.000124
2016-04-06 02:17:26    0.000124
2016-04-06 14:17:04    0.000124
2016-04-06 19:16:38    0.000124
2016-04-06 05:16:14    0.000124
2016-04-06 15:16:45    0.000124
2016-04-06 08:44:19    0.000124
2016-04-06 22:17:26    0.000124
2016-04-07 05:16:17    0.000124
2016-04-06 20:48:27    0.000124
2016-04-07 06:46:12    0.000124
2016-04-06 10:17:12    0.000124
2016-04-06 02:16:12    0.000124
2016-04-07 03:45:23    0.000124
2016-04-06 15:45:50    0.000124
2016-04-06 14:44:55    0.000124
2016-04-06 13:17:03    0.000124
2016-04-07 04:46:51    0.000124
2016-04-05 17:26:16    0.000103
2016-04-05 20:47:00    0.000103
2016-04-05 20:47:37    0.000103
2016-04-07 10:17:20    0.000103
2016-04-06 01:47:07    0.000103
2016-04-06 03:16:25    0.000103
                         ...   
2016-03-

In [57]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [58]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [59]:
autos['last_seen']

0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
2        2016-04-06 20:15:37
3        2016-03-15 03:16:28
4        2016-04-01 14:38:50
5        2016-04-06 09:45:21
6        2016-03-23 02:48:59
7        2016-04-07 03:17:32
8        2016-03-26 18:18:10
9        2016-04-06 10:46:35
10       2016-04-07 03:16:21
11       2016-03-16 18:45:34
12       2016-04-06 14:17:52
13       2016-04-01 14:17:13
14       2016-04-01 15:46:47
15       2016-04-02 21:10:48
16       2016-03-18 05:29:37
17       2016-04-05 20:46:26
18       2016-04-02 12:16:41
19       2016-03-17 14:45:58
20       2016-03-07 05:46:46
21       2016-04-05 23:46:19
22       2016-04-01 06:45:30
23       2016-03-17 08:45:17
24       2016-04-07 13:16:50
25       2016-04-01 05:16:49
26       2016-04-05 23:46:48
28       2016-04-07 14:58:48
29       2016-04-02 12:45:44
30       2016-03-14 11:47:31
                ...         
49968    2016-04-05 13:16:35
49969    2016-03-25 23:18:15
49970    2016-04-07 04:17:34
49971    2016-

In [60]:
autos["last_seen"].value_counts(normalize=True,dropna= False)

2016-04-07 06:17:27    0.000165
2016-04-07 03:16:17    0.000144
2016-04-06 21:17:51    0.000144
2016-04-06 06:17:24    0.000124
2016-04-06 07:46:11    0.000124
2016-04-05 16:44:47    0.000124
2016-04-06 15:17:56    0.000124
2016-04-06 02:17:26    0.000124
2016-04-06 14:17:04    0.000124
2016-04-06 19:16:38    0.000124
2016-04-06 05:16:14    0.000124
2016-04-06 15:16:45    0.000124
2016-04-06 08:44:19    0.000124
2016-04-06 22:17:26    0.000124
2016-04-07 05:16:17    0.000124
2016-04-06 20:48:27    0.000124
2016-04-07 06:46:12    0.000124
2016-04-06 10:17:12    0.000124
2016-04-06 02:16:12    0.000124
2016-04-07 03:45:23    0.000124
2016-04-06 15:45:50    0.000124
2016-04-06 14:44:55    0.000124
2016-04-06 13:17:03    0.000124
2016-04-07 04:46:51    0.000124
2016-04-05 17:26:16    0.000103
2016-04-05 20:47:00    0.000103
2016-04-05 20:47:37    0.000103
2016-04-07 10:17:20    0.000103
2016-04-06 01:47:07    0.000103
2016-04-06 03:16:25    0.000103
                         ...   
2016-03-

In [61]:
autos["last_seen"].sort_index()

0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
2        2016-04-06 20:15:37
3        2016-03-15 03:16:28
4        2016-04-01 14:38:50
5        2016-04-06 09:45:21
6        2016-03-23 02:48:59
7        2016-04-07 03:17:32
8        2016-03-26 18:18:10
9        2016-04-06 10:46:35
10       2016-04-07 03:16:21
11       2016-03-16 18:45:34
12       2016-04-06 14:17:52
13       2016-04-01 14:17:13
14       2016-04-01 15:46:47
15       2016-04-02 21:10:48
16       2016-03-18 05:29:37
17       2016-04-05 20:46:26
18       2016-04-02 12:16:41
19       2016-03-17 14:45:58
20       2016-03-07 05:46:46
21       2016-04-05 23:46:19
22       2016-04-01 06:45:30
23       2016-03-17 08:45:17
24       2016-04-07 13:16:50
25       2016-04-01 05:16:49
26       2016-04-05 23:46:48
28       2016-04-07 14:58:48
29       2016-04-02 12:45:44
30       2016-03-14 11:47:31
                ...         
49968    2016-04-05 13:16:35
49969    2016-03-25 23:18:15
49970    2016-04-07 04:17:34
49971    2016-

In [62]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [63]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

In [64]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

###### odd values min = 1000 and max = 9999,the registration falls in between 1999 upper quartile , 2003 in 50% and 2008 in lower quartile
###### Dealing with Incorrect Registration Year Data

In [65]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

In [66]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

Exploring Price by Brand

In [77]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

In [81]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [82]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

Exploring Mileage

In [83]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_mileage"])

mean_mileage
audi                   9336
bmw                    8332
ford                   3749
mercedes_benz          8628
opel                   2975
volkswagen             5402

In [84]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [85]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [68]:
autos[~autos["year_of_registration"].between(1990,2020)]


KeyError: 'year_of_registration'

In [69]:
autos["year_of_registration"].value_counts(normalize = True)

KeyError: 'year_of_registration'

In [70]:
autos["year_of_registration"].value_counts().head(50)

KeyError: 'year_of_registration'

In [71]:
autos['year_of_registration'].quantile([0.1, 0.5, 0.9])

KeyError: 'year_of_registration'

In [72]:
min = 1996
max = 2013
# 

In [73]:
plt.figure(figsize=(10, 6))
min_ = autos['year_of_registration'].quantile(0.1)
max_ = autos['year_of_registration'].quantile(0.99)
price_data = autos.loc[autos['year_of_registration'].between(min_, max_), 'year_of_registration']
ax = sns.violinplot(price_data)
ax.set_title("Auto[year_of_registration] Violinplot with min:{} and max:{}".format(min_, max_))

NameError: name 'plt' is not defined

In [74]:
common_brands= autos["brand"].unique()

In [75]:
brand_mean_prices = {}
for b in common_brands:
    brand_only = autos.loc[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
    

NameError: name 'brand' is not defined

In [76]:
top_brands = autos["brand"].value_counts().index[:21]
top_brands_mean_price = {}
for brand in top_brands: 
    brand_only = autos[autos["brand" == brand]] 
    mean_price = brand_only["price"].mean() 
    top_brand_mean_price[brand] = int(mean_price)

KeyError: False

In [85]:
brand_mean_prices = {}
def float_to_int(x):
    if x == float('inf') or x == float('-inf'):
        return float('nan') # or a large value you choose
    return int(x)


In [87]:
for brand in common_brands: 
    brand_only = autos[autos["brand"] == brand] 
    mean_price = brand_only["price"].mean() 
    brand_mean_prices[brand] = float(mean_price)



In [ ]:
autos["price"] = autos[autos["brand"]==brand]